core/templates/services/playthrough.service.spec.ts

```diff
@@ -298,6 +325,99 @@ describe('PlaythroughService', () => {
       });
     });
 
+    it('should identify disjoint cycle rotations', () => {
+      spyOn(stopwatchObjectFactory, 'create').and.returnValue(
+        jasmine.createSpyObj('Stopwatch', {getTimeInSecs: 370, reset: null}));
+
+      playthroughService.recordExplorationStartAction('stateName1');
```

Comment:
I would advise having more, smaller tests that verify specific facets / cases -- and then one big one, like the example I gave perhaps, as an "integration test". Cf. earlier discussion today about being careful to probe the various cases more thoroughly to avoid unexpected errors in production.

For each test, add a comment describing what we expect cycleOccurrences to be and explain why it should be that way.

Reply:
I'm pretty confident in the test coverage actually, I've been adding them for each key behavior I've introduced. Is there a specific hole in coverage that's missing?

We won't be able to write integration tests until the playthroughs are actually rendered, given we've recently removed the playthroughs directive from the stats tab.

Reply:
By "integration tests" I don't mean e2e tests. I just mean longer chains of stuff like you have here.

For smaller tests I was thinking about writing the minimal test that would check specific properties of the algorithm that you want to assert, and this test doesn't really seem minimal (unless I'm missing something).

Reply:
I think there's still a misunderstanding here. I'm not talking about testing the individual functions at all.

I'm talking about testing the logic of cycle detection. There are certain properties you want the cycle detection algorithm to satisfy. I was suggesting having minimal test cases for each of these properties (in terms of the lengths of the playthroughs used to test them), as well as a couple of additional longer composite playthroughs. Note that I meant "properties" in the layperson sense, not in the technical "attributes of a class" sense. I hope this clarifies!

Reply:
This has been addressed in the newest changes, along with a reorganized test structure + helper methods, and a refactored cycle detection algorithm (implemented as we had discussed offline).